tdclient/connector_api.py

```diff
@@ -16,12 +16,70 @@ def connector_guess(self, job):
 
         Args:
             job (dict): :class:`dict` representation of `seed.yml`
+                        See Also: https://www.embulk.org/docs/built-in.html#guess-executor
 
         Returns:
-             :class:`dict`
+             :class:`dict`: The configuration of the Data Connector.
+
+        Examples:
+            >>> config = {
+            ...     "in": {
+            ...         "type": "s3",
+            ...         "bucket": "your-bucket",
+            ...         "path_prefix": "logs/csv-",
+            ...         "access_key_id": "YOUR-AWS-ACCESS-KEY",
+            ...         "secret_access_key": "YOUR-AWS-SECRET-KEY"
+            ...     },
+            ...     "out": {"mode": "append"},
+            ...     "exec": {"guess_plugins": ["json", "query_string"]},
+            ... }
+            >>> td.api.connector_guess(config)
+            {'config': {'in': {'type': 's3',
+               'bucket': 'your-bucket',
+               'path_prefix': 'logs/csv-',
+               'access_key_id': 'YOUR-AWS-ACCESS-KEY',
+               'secret_access_key': 'YOU-AWS-SECRET-KEY',
+               'parser': {'charset': 'UTF-8',
+                'newline': 'LF',
+                'type': 'csv',
+                'delimiter': ',',
+                'quote': '"',
+                'escape': '"',
+                'trim_if_not_quoted': False,
+                'skip_header_lines': 1,
+                'allow_extra_columns': False,
+                'allow_optional_columns': False,
+                'columns': [{'name': 'sepal.length', 'type': 'double'},
+                 {'name': 'sepal.width', 'type': 'double'},
+                 {'name': 'petal.length', 'type': 'double'},
+                 {'name': 'petal.width', 'type': 'string'},
+                 {'name': 'variety', 'type': 'string'}]}},
+              'out': {'mode': 'append'},
+              'exec': {'guess_plugin': ['json', 'query_string']},
+              'filters': [{'rules': [{'rule': 'upper_to_lower'},
+                 {'pass_types': ['a-z', '0-9'],
+                  'pass_characters': '_',
+                  'replace': '_',
+                  'rule': 'character_types'},
+                 {'pass_types': ['a-z'],
+                  'pass_characters': '_',
+                  'prefix': '_',
+                  'rule': 'first_character_types'},
+                 {'rule': 'unique_number_suffix', 'max_length': 128}],
+                'type': 'rename'},
+               {'from_value': {'mode': 'upload_time'},
+                'to_column': {'name': 'time'},
+                'type': 'add_time'}]}}
         """
         headers = {"content-type": "application/json; charset=utf-8"}
-        payload = json.dumps(job).encode("utf-8") if isinstance(job, dict) else job
+        if isinstance(job, dict):
+            if "config" not in job:
+                job = {"config": job}
```

Comment:
Did you consider raising an exception when a dictionary `job` doesn't have the `config` key?

Always wrapping job by `config` might make debugging harder. For example, if there is a typo `confg`, the dictionary automatically becomes:

```json
{
  "config":
    "confg": {
      "in": {
      ...
```

Reply:
This is similar to what TD cli does https://github.com/treasure-data/td/blob/15495f12d8645a7b3f6804098f8f8aca72de90b9/lib/td/command/connector.rb#L70
I don't think that API behavior wasn't described explicitly before (even if @kiyoto couldn't use this function!), so I don't think we don't let users to care about config key itself.

Reply:
Fine, then can we clone the complete normalization logic? https://github.com/treasure-data/td/blob/15495f12d8645a7b3f6804098f8f8aca72de90b9/lib/td/connector_config_normalizer.rb#L7-L30

Reply:
ok, let me fix to have normalized config function.

Reply:
@takuti fixed on 94f4da0